File: 3.无重复字符的最长子串.c
/*
 * @lc app=leetcode.cn id=3 lang=c
 *
 * [3] 无重复字符的最长子串
 */

// @lc code=start
int lengthOfLongestSubstring(char* s)
{
    int n = strlen(s);
    int map[128] = { 0 }; // 用于记录字符最后出现的位置，类似于哈希表
    int left = 0, right = 0;
    int maxLength = 0;

    while (right < n) {
        char currentChar = s[right]; // 当前字符
        if (map[currentChar] > 0) { // 如果当前字符已经出现过
            left = (map[currentChar] > left) ? map[currentChar] : left; // 如果当前字符上次出现的位置在left右边，则更新left
        }
        map[currentChar] = right + 1; // 更新字符最后出现的位置
        maxLength = (maxLength > (right - left + 1)) ? maxLength : (right - left + 1);
        right++;
    }

    return maxLength;
}
// @lc code=end

File: 3.无重复字符的最长子串.py
#
# @lc app=leetcode.cn id=3 lang=python3
#
# [3] 无重复字符的最长子串
#


# @lc code=start
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        """解题思路：
        1. 使用左右指针 left 和 right 表示当前窗口
        2. 用哈希表记录每个字符最后出现的位置
        3. 右指针不断向右移动扩大窗口
        4. 当遇到重复字符时，将左指针移动到重复字符上次出现位置的下一位
        5. 过程中更新最大窗口长度
        """
        char_to_index = {}
        left = 0
        max_length = 0
        for right in range(len(s)):
            if s[right] in char_to_index:
                left = max(left, char_to_index[s[right]] + 1)
            char_to_index[s[right]] = right
            max_length = max(max_length, right - left + 1)
        return max_length


# @lc code=end
